src/properties.c

```diff
@@ -0,0 +1,550 @@
+
+	return 0;
+}
+
+static int set_data_items(struct knot_thing *thing, int fd)
+{
+	int rc;
+	int i;
+	char **data_item_group;
+
+	int sensor_id;
+	int reg_addr;
+	int bit_offset;
+	knot_schema schema;
+	knot_config config;
+
+	data_item_group = get_data_item_groups(fd);
+
+	for (i = 0; data_item_group[i] != NULL ; i++) {
+		rc = set_sensor_id(thing, fd, data_item_group[i], i,
+				   &sensor_id);
+		if (rc < 0)
+			goto error;
+
+		rc = set_schema(thing, fd, data_item_group[i], &schema);
+		if (rc < 0)
+			goto error;
+
+		rc = set_config(thing, fd, data_item_group[i], schema, &config);
+		if (rc < 0)
+			goto error;
+
+		rc = set_modbus_source_properties(thing, fd,
+						  data_item_group[i], schema,
+						  &reg_addr, &bit_offset);
+		if (rc < 0)
+			goto error;
+
+		device_set_new_data_item(thing, sensor_id, schema, config,
+					 reg_addr, bit_offset);
+	}
+
+	l_strfreev(data_item_group);
+
+	return 0;
+
+error:
+	l_strfreev(data_item_group);
+
+	return -EINVAL;
+}
+
+static int set_thing_name(struct knot_thing *thing, int fd)
+{
+	char *knot_thing_name;
+
+	knot_thing_name = storage_read_key_string(fd, THING_GROUP, THING_NAME);
+	if (knot_thing_name == NULL || !strcmp(knot_thing_name, "") ||
+	    strlen(knot_thing_name) >= KNOT_PROTOCOL_DEVICE_NAME_LEN) {
+		l_free(knot_thing_name);
+		return -EINVAL;
+	}
+
+	device_set_thing_name(thing, knot_thing_name);
+	l_free(knot_thing_name);
+
+	return 0;
+}
+
+static int set_thing_user_token(struct knot_thing *thing, int fd)
+{
+	char *user_token;
+
+	user_token = storage_read_key_string(fd, THING_GROUP, THING_USER_TOKEN);
+	if (user_token == NULL || !strcmp(user_token, "")) {
+		l_free(user_token);
+		return -EINVAL;
+	}
+	device_set_thing_user_token(thing, user_token);
+
+	return 0;
+}
+
+static int set_thing_credentials(struct knot_thing *thing, char *filename)
+{
+	int cred_fd;
+	char *thing_id;
+	char *thing_token;
+
+	cred_fd = storage_open(filename);
+	if (cred_fd < 0)
+		return cred_fd;
+
+	thing_id = storage_read_key_string(cred_fd, CREDENTIALS_GROUP,
+					  CREDENTIALS_THING_ID);
+
+	thing_token = storage_read_key_string(cred_fd, CREDENTIALS_GROUP,
+					     CREDENTIALS_THING_TOKEN);
+
+	device_set_thing_credentials(thing, thing_id, thing_token);
+
+	l_free(thing_id);
+	l_free(thing_token);
+
+	storage_close(cred_fd);
+
+	return 0;
+}
+
+static int set_thing_properties(struct knot_thing *thing, char *filename)
+{
+	int device_fd;
+	int rc;
+
+	device_fd = storage_open(filename);
+	if (device_fd < 0)
+		return device_fd;
+
+	rc = set_thing_name(thing, device_fd);
+	if (rc < 0)
+		return rc;
+
+	rc = set_thing_user_token(thing, device_fd);
+	if (rc < 0)
+		return rc;
+
+	rc = set_modbus_slave_properties(thing, device_fd);
+	if (rc < 0)
+		return rc;
+
+	rc = set_data_items(thing, device_fd);
+	if (rc < 0)
+		return rc;
+
+	storage_close(device_fd);
+
+	return rc;
+}
+
+int properties_create_device(struct knot_thing *thing,
+			     struct device_settings *conf_files)
```

Comment:
Can you check if this column alignment is correct in your text editor?

Reply:
It is correct in the editor, it must be a GitHub bug.